### ISSUE TITLE
[Feature] phashing support

### ISSUE THREAD

**Issue by boi123212321** [boi123212321/porn-manager#503]

**Comment by gramseyBES** [comment]
Not sure if this helps, but I'd worked up a p-hashing approach that you can reference/start from.  It uses imagemagick though, as I'd intended this to run on a Linux server.  I hadn't gotten this working for videos yet, but I imagine it would work pretty well with frame-grabs from a video and hashing those.

Call to get an image's pixels and hash the results:
https://github.com/Black-Eagle-Software/bes_CMS/blob/f0be4cc24e7129668ac24aef11ff805a56f4aace/src/routes/media/socket-upload-process.js#L58-L73

getImagePixels calls out to imagemagick to resize the image and convert to grayscale:
https://github.com/Black-Eagle-Software/bes_CMS/blob/master/src/helpers/getImagePixels.js

perceptualHashing takes the output from getImagePixels and performs the transforms to generate the hash:
https://github.com/Black-Eagle-Software/bes_CMS/blob/master/src/helpers/perceptualHashing.js

**Comment by boi123212321** [comment]
@gramseyBES Actually it's pretty easy to do with the modules I already have installed. Take a screenshot at 50% time with ffmpeg, then use JIMP to make it grayscale and get the hash. Store in DB.

**Comment by boi123212321** [comment]
Blocked by https://github.com/porn-vault/porn-vault/issues/945

**Comment by CPSibo** [comment]
Idk if it helps but I had to implement pHashing for a Laravel-Vue project. You might find it easier to just write the pHashing manually and not have a dependency. The only dependency in my code below is `exif-js`, but only for one function, so it could probably be replaced.

My code was intended be doing duplicate and "similar image" detection across a pretty large database of images, so I wanted to be able to do a pre-check on the client side before wasting the server's time with duplicate uploads. This meant porting a pHashing algorithm from PHP to JS. In the end, I got the client to produce identical pHashes as my PHP library, enabling some pretty cool stuff on the client-side.

For the backend, I used [imagehash](https://github.com/jenssegers/imagehash)'s "Difference Hash". Then converted its hex output into decimal with a simple `base_convert(..., 16, 10)` and stored that in a BigInt field in the database. This would be a pretty "standard" perceptual hash for a still image. And I'm only storing the data as an integer because of MariaDB quirks; you could just as easily skip the parts where I convert to decimal and store things in binary or hex.

Doing all that means I can get image distances really quickly on the server side by doing `BIT_COUNT(phash ^ phashToCheck)` in MySQL, where `phash` is the pHash column in the DB and `phashToCheck` is the pHash that I'm passing in from the potential upload. So, BIT_COUNT() is just checking how many bits in the argument are `1` and the `^` is doing a bit-wise XOR of the two pHashes. Thus, we're counting how many bits are different between the two hashes, giving us a really easy distance metric. This should be easy enough to emulate in JS.

Then, I needed to make Vue produce the same pHash as PHP would, if the image was actually uploaded. The resulting code is actually fairly short:
```js
  // `imageObj` is a vanilla JS `Image` object. 
  getPhash = (imageObj) => {
    const imageWidth = imageObj.width
    const imageHeight = imageObj.height

    // Getting an `Image`'s data is easiest when we just get it from a canvas.
    const canvas = document.createElement('canvas')
    const context = canvas.getContext('2d')
    context.drawImage(imageObj, 0, 0)
    const imageData = context.getImageData(0, 0, imageWidth, imageHeight)

    // `pixels` will store our image's pixel data as a 2D array.
    const pixels = this.createArray(imageHeight, imageWidth)
    const hash = []

    // Step through our image's data, assuming 16-bit color depth (r, g, b, and a).
    for (let i = 0; i < imageData.data.length; i += 4) {
      // Translate our 1D loop index into 2D indices.
      const j = (i === 0) ? 0 : i / 4
      const x = j % imageWidth
      const y = Math.floor((j / imageWidth) % imageHeight)

      // Get each channel's value. Ignore alpha.
      const r = imageData.data[i]
      const g = imageData.data[i + 1]
      const b = imageData.data[i + 2]

      // Get a perceptual brightness value for the pixel based on this widely-used magic 
      // formula that is supposed to represent human sensitivity to each color.
      const gs = Math.floor((r * 0.299) + (g * 0.587) + (b * 0.114))

      // Stash our pixel value in the array.
      pixels[y][x] = gs
    }

    for (let y = 0; y < imageHeight; y++) {
      let left = pixels[y][0]

      for (let x = 1; x < imageWidth; x++) {
        const right = pixels[y][x]

        // The "difference" hash stores whether the preceding pixel was brighter or darker
        // than the current pixel. 
        hash.push(left > right ? 1 : 0)

        // Prepare the next loop.
        left = right
      }
    }

    // Turn the array into a string representing our binary number.
    const binaryHash = hash.join('')

    // Finally, turn our binary number into a decimal number, since that's how we store pHashes in the DB.
    return parseInt(binaryHash, 2)
  }

  // Recursive function to create arbitrarily-sized arrays.
  createArray = (length, ...args) => {
    const arr = new Array(length || 0)
    let i = length

    if (args.length > 0) {
      while (i--) { arr[length - 1 - i] = this.createArray(...args) }
    }

    return arr
  }
```

The only thing missing here is that most pHashing algorithms use very small versions of the original image. For my PHP dependency, it used 9x8 px versions of the original. So I also do that in JS with another function before passing the `Image` object into `getPhash()`. This uses the canvas, again, to make transforming the image easier: [ImageTools.js](https://gist.github.com/CPSibo/723bb6a8da88b6c219e89d7491bc178c)

The relevant bit there is `resizeIgnoreAspect()`. It takes in a vanilla JS `File` object (like you'd get from a standard HTML form), and an object like `{ width: 9, height: 8 }` to specify the desired dimensions. So I use it like:
```js
// Again, `file` here is a vanilla JS `File` object.
const resizedFile = await this.imageTools.resizeIgnoreAspect(file, { width: 9, height: 8 })

const imageObj = new Image()
imageObj.addEventListener('load', () => {
  const phash = this.getPhash(imageObj)

  urlCreator.revokeObjectURL(imageUrl)

  // This actually resides in a Promise, for me, so return the data however you want.
  resolve(phash)
}, false)

const urlCreator = window.URL || window.webkitURL
const imageUrl = urlCreator.createObjectURL(resizedFile)
imageObj.src = imageUrl
```

So, when it's all said and done, I just use the snippet above to get the pHash of the client's image then hand the integer off to my API to get any images where the distance metric is < some threshold. I was having decent results around 15 to 20 for the threshold. And it's pretty snappy (not that I tested it with any video uploads).

For your case, you can probably replace my `File` objects with something suitable to read the output of ffmpeg.

It seems like a lot of cruft but it's kind of simple code for how powerful it is. Having a native solution in JS lets you do all kinds of stuff in Vue, right where the user is uploading something or searching for something. I also have a native MD5 hashing solution for images, to detect exact dupes, if you need.

**Comment by boi123212321** [comment]
@CPSibo I have already tried implementing blockhash myself (using imagemagick), and it kinda works, but the problem is I don't get hash collisions that often for images that are the same, but in different quality. So instead, I could do distance checks, but I'm still figuring out how to do it efficiently, without having to check every hash each time. A BK tree would work, but I haven't seen an implementation that allowed for items to be spliced out of it, yet. I could just do brute force, and to be honest, that might be the thing i'm going for for now, as it shouldn't make any difference in how the rest of the hashing is implemented.

**Comment by CPSibo** [comment]
Tbh, doing a bit-wise XOR on two ints is pretty damn quick. The algorithm I outlined above worked well for my stress test data of several million images. I imagine most people will have far fewer videos in their library, here.

If you really needed the speed, you could cache the phash:id in something like redis, but this project's docker already uses a sizable amount of RAM. Caching the phashes would probably bump the usage up another few hundred MB.

Alternatively, you could always defer the check until later. For instance, if you're planning to use this feature to help detect dups that are loaded from disk automatically, you could provide a simple page where the user could view potential duplicates and acts on them (ignore, delete, etc), similar to how hydrus does it. Then you could populate that page in the background, rather than in real-time.

In any case, you might be preemptively optimizing something that won't be a real performance problem in the majority of users' cases.

**Comment by boi123212321** [comment]
Yeah I think it should be fine either way, it will probably support different hashes anyway, so I'll go for blockhash as a start. But first, I need to get imagemagick (or alternative) implemented
